Update aws lambda spec to remove X-Ray Env propagation (#3166)
open-telemetry/opentelemetry-specification: specification/trace/semantic_conventions/instrumentation/aws-lambda.md

[faasres]: ../../../resource/semantic_conventions/faas.md (FaaS resource conventions)
[cloud]: ../../../resource/semantic_conventions/cloud.md (Cloud resource conventions)

### AWS X-Ray Environment Span Link

If the `_X_AMZN_TRACE_ID` environment variable is set, instrumentation SHOULD try to parse an
OpenTelemetry `Context` out of it using the [AWS X-Ray Propagator](../../../context/api-propagators.md). If the

> resulting `Context` is [valid](../../api.md#isvalid) then a [Span Link][] should be added to the new Span's

Reviewer: Is this meant as a normative statement? i.e.

```suggestion
resulting `Context` is [valid](../../api.md#isvalid) then a [Span Link][] SHOULD be added to the new Span's
```

> [start options](../../api.md#specifying-links) with an associated attribute of `source=x-ray-env` to

Reviewer 2: Maybe just `source=x-ray`?

Author: I didn't want to confuse it with regular x-ray header propagation.

Reviewer 2: I now understand that it's for env var, so I'll consider `x-ray-env-var` (but both OK IMO)

Author: I'm going to go with `x-ray-env`.  If you have a strong opinion otherwise, please open another PR.

indicate the source of the linked span.

> Instrumentation needs to check if the context is valid because the `_X_AMZN_TRACE_ID` environment variable may

Reviewer: ```suggestion
Instrumentation MUST check if the context is valid before using it because the `_X_AMZN_TRACE_ID` environment variable can
```

[faasres]: ../../../resource/semantic_conventions/faas.md (FaaS resource conventions)
[cloud]: ../../../resource/semantic_conventions/cloud.md (Cloud resource conventions)

### EventToCarrier

> Each event type transmits propagation data inconsistently. Instrumentation SHOULD define a function that
> takes the event as an argument and returns a carrier that can be passed into the configured propagators.

Reviewer 3: ```suggestion
Each event type may represent propagation data in its own way. Instrumentation SHOULD accept a callback function that
takes the event as an argument and returns a carrier that can be passed into the configured propagators.
```

### AWS X-Ray Environment Span Link

If the `_X_AMZN_TRACE_ID` environment variable is set, instrumentation SHOULD try to parse an
OpenTelemetry `Context` out of it using the [AWS X-Ray Propagator](../../../context/api-propagators.md). If the

> resulting `Context` is [valid](../../api.md#isvalid) then a span link should be added to the lambda span with

Reviewer 4: ```suggestion
resulting `Context` is [valid](../../api.md#isvalid) then a [Span Link](https://opentelemetry.io/docs/concepts/signals/traces/#span-links) should be added to the new Span's [start options](../../api.md#specifying-links) with
```

> an associated attribute of `source=x-ray-env` to indicate what the link is for.

Reviewer 5: ```suggestion
an associated attribute of `source=x-ray-env` to indicate the source of the linked span.
```

> an associated attribute of `source=x-ray-env` to indicate what the link is for.

Reviewer 4: ```suggestion
an associated attribute of `source=x-ray-env` to indicate what the Link is for.
```

> We check if it is valid because sometimes the `_X_AMZN_TRACE_ID` environment variable contains

Reviewer 5: ```suggestion
Instrumentation needs to check if it the context is valid because the `_X_AMZN_TRACE_ID` environment variable may contain
```

an incomplete trace context which indicates X-Ray isn’t enabled. The environment variable will be set and the
`Context` will be valid and sampled only if AWS X-Ray has been enabled for the Lambda function. A user can

> disable AWS X-Ray for the function if the X-Ray span link is not desired.

Reviewer 5: Is there a need to recommend a mechanism to disable span links when X-Ray is enabled?

Reviewer 4: A link should be lightweight enough to not require a way to disable outside of disabling x-ray.

> disable AWS X-Ray for the function if the X-Ray span link is not desired.

Reviewer 4: ```suggestion
disable AWS X-Ray for the function if the X-Ray Span Link is not desired.
```

[faasres]: ../../../resource/semantic_conventions/faas.md (FaaS resource conventions)
[cloud]: ../../../resource/semantic_conventions/cloud.md (Cloud resource conventions)

### EventToCarrier

Each event type may represent propagation data in its own way. Instrumentation SHOULD support a user defined transform
function which takes the event as an argument and returns a carrier that can be passed into the configured propagators.
Instrumentation SHOULD provide default implementations for the standard event types.

### AWS X-Ray Environment Span Link

> If the `_X_AMZN_TRACE_ID` environment variable is set, instrumentation SHOULD try to parse an

Reviewer 4: Should this be MUST?

Author: I'd rather not make it required, but @Aneurysm9 what do you think?

Reviewer 3: Let's leave it as `SHOULD` for now and we can discuss separately whether to change to `MUST`.  I think `SHOULD` is probably fine and we should scowl menacingly at any OTel project that doesn't do this.
